parser.py

```diff
@@ -4,42 +4,132 @@
 from utils import *
 from lxml import etree
 
-class State:
-    class Waiting: pass
-    class Meta: pass
-    class Ingredients: pass
-    class Steps: pass
+
+class Line:
+    """Holds the type and contents of a line, e.g. an ingredient or an instruction.
+
+    In case that word wrapping was used in the original markup, one Line class can
+    actually hold the contents of multiple lines in the source code belonging to the
+    same ingredient, instruction etc.
+    """
+    class Empty: pass
+    class Plain: pass
+    class Comment: pass
+    class Command: pass
+    class Part: pass
+    class Ingredient: pass
+    class Step: pass
+    class Note: pass
     class WaitPhase: pass
-    class FinishPhase: pass
-    class FinishRecipe: pass
 
-def parseIngredient(line):
-    r = r'^\s*#\s*([0-9/\.]+(?:\s+[0-9/]+)?)([a-zA-Z]+)?\s+(.*)'
-    result = re.match( r, line)
+    startCharacters = (
+        ("'", Comment),
+        ('!', Command),
+        ('#', Part),
+        ('*', Ingredient),
+        ('--', Note), # notes have to be parsed before steps!
+        ('-', Step),
+        ('+', WaitPhase),
+    )
+
+    def __init__(self, lineType = None, lineNo = None, contents = None):
+        self.lineType = lineType
+        self.lineNo = lineNo
+        self.contents = contents
+
+    def __repr__(self):
+        return 'Line({!r}, {!r}, {!r})'.format(self.lineType, self.lineNo, self.contents)
+
+    @property
+    def contents(self):
+        if isinstance(self._contents, str):
+            return self._contents
+        else:
+            return ' '.join(self._contents)
 
-    if result is None:
-        line = line.lstrip('# \t')
-        line = line.rstrip(' \t\n')
-        return Ingredient( line, None, None)
+    @contents.setter
+    def contents(self, contents):
+        self._contents = contents
+
+    def append(self, contents):
+        if isinstance(self._contents, str):
```

Comment:
Then there is often an unnecessary ' '.join(), but I agree that this is the better solution and will change it.